src/multivariate.jl

```diff
@@ -151,65 +171,92 @@ function _logpdf(d::TuringMvLogNormal, x::AbstractVector{T}) where {T<:Real}
     return insupport(d, x) ? (_logpdf(d.normal, log.(x)) - sum(log.(x))) : -T(Inf)
 end
 function _logpdf(d::TuringMvLogNormal, x::AbstractMatrix{T}) where {T<:Real}
-    if all(insupport(d, x))
+    insupp = map(1:size(x, 2)) do i 
+        insupport(d, x[:, i])
+    end
+    if all(insupp)
         return _logpdf(d.normal, log.(x)) - vec(sum(log.(x), dims=1))
     else
-        return fill(-T(Inf), size(x, 2))
+        return map(1:size(x, 2)) do i
+            insupp[i] && return -T(Inf)
+            return _logpdf(d.normal, log.(x[:,i])) - sum(log.(x[:,i]))
```

Comment:
I guess that could be simplified to just `map(_logpdf(d, view(x, :, i)), 1:size(x, 2))` or something similar.

Reply:
Hmm this will be computationally more expensive, but it solves the type stability.

Reply:
Actually it doesn't work because we need to take `log` inside which fails when `x` is not in the support. So we have to return `-Inf` separately.

Reply:
I don't see why that would not work if a column of `x` is not in the support - we would just call
https://github.com/TuringLang/DistributionsAD.jl/blob/2ac40ba38f78662e57a91399bac2b7f711cac4d3/src/multivariate.jl#L170-L172
for every column. So as long as the logpdf definition for vectors is type-stable, the matrix-valued version should be trivial? Or am I missing something here?

So IMO the type-stability issue should be handled in the implementation for vectors, and then the definition for matrices could be a simple map.

Reply:
Oh I see, ok. Yes that makes sense. There are still 2 issues though:
1. `insupport` would be called twice, once in the matrix version to find `insupp` and another in the vector version.
2. The vector version has the same `T(Inf)` that may not be type stable or even worse may error if `T` is `Int`.

I think the current implementation avoids the double calling of `insupport` and handles the `Int` case and type stability nicely.

Reply:
That's not exactly what I had in mind. I thought you could just define
```julia
function _logpdf(d::TuringMvLogNormal, x::AbstractMatrix{<:Real}) 
    return map(i -> _logpdf(d, view(x, :, i)), axes(x, 2))
end
```
and move any potential fix for the type instability to the definition of `_logpdf(::TuringMvLogNormal, x::AbstractVector{<:Real})`. As long as the implementation for vectors is type stable, there is no type instability, and additionally there is not double calling of `insupport`.

Reply:
This will be slower than the matrix version when all columns are in the support and it allocates for the views.

Reply:
Well view allocation is still better than `x[:,i]`.

Reply:
```julia
julia> dist = TuringMvLogNormal(MvNormal(rand(10), rand(10)));

julia> x = rand(dist, 100);

julia> f(dist, x) = logpdf(dist, x);

julia> g(dist, x) = [logpdf(dist, view(x, :, i)) for i in 1:size(x,2)];

julia> @btime f($dist, $x);
  18.800 μs (17 allocations: 29.63 KiB)

julia> @btime g($dist, $x);
  35.700 μs (503 allocations: 68.13 KiB)
```

Reply:
I will mark this as resolved. Feel free to unmark it if you think there is a better way to do this.

Reply:
While performing some benchmarks I noticed that the current implementation is incorrect - it throws an error if samples are not in the support due to the line `insupp[i] && return -T(Inf)` which should be `insupp[i] || return -T(Inf)`.

I performed the following benchmarks (after fixing this bug but without fixing any type instabilities in `_logpdf(d::TuringMvLogNormal, x::AbstractVector)` since for the benchmarked types there should not be any issues):
```julia
using DistributionsAD
using BenchmarkTools

dist = TuringMvLogNormal(TuringDiagMvNormal(rand(10), rand(10)))

x = rand(dist, 100)
y = copy(x)
y[3, 20] = -1

f(dist, x) = DistributionsAD._logpdf(dist, x)
g(dist, x) = [DistributionsAD._logpdf(dist, view(x, :, i)) for i in axes(x, 2)]
function h(dist, x)
    if all(i -> DistributionsAD.insupport(dist, view(x, :, i)), axes(x, 2))
        logx = log.(x)
        return DistributionsAD._logpdf(dist.normal, logx) - vec(sum(logx; dims = 1))
    else
        return [DistributionsAD._logpdf(dist, view(x, :, i)) for i in axes(x, 2)]
    end
end

@btime f($dist, $x) #   20.266 μs (17 allocations: 29.63 KiB)
@btime g($dist, $x) #   37.591 μs (503 allocations: 68.13 KiB)
@btime h($dist, $x) #   11.583 μs (14 allocations: 21.47 KiB)

@btime f($dist, $y) #   37.841 μs (702 allocations: 73.97 KiB)
@btime g($dist, $y) #   36.997 μs (499 allocations: 67.50 KiB)
@btime h($dist, $y) #   37.237 μs (500 allocations: 67.53 KiB)
```

BTW, benchmarking variant `h2` of `h` that does not precompute `logx = log.(x)`
```julia
@btime h2($dist, $x) #   19.859 μs (15 allocations: 29.41 KiB)
@btime h2($dist, $y) #   37.209 μs (500 allocations: 67.53 KiB)
```
After changing
```julia
function _logpdf(d::TuringMvLogNormal, x::AbstractVector{T}) where {T<:Real}
    return insupport(d, x) ? (logx = log.(x); _logpdf(d.normal, logx) - sum(logx)) : -T(Inf)
end
```
I get
```julia
@btime g($dist, $x) #   26.275 μs (403 allocations: 52.50 KiB)

@btime g($dist, $y) #   26.049 μs (400 allocations: 52.03 KiB)
@btime h($dist, $y) #   26.614 μs (401 allocations: 52.06 KiB)
```

Overall, I think it is more important to improve performance in the case where all samples are in the support of the distribution, since that should be the more common case. IMO the benchmarks suggest implementing alternative `h` and pre-computing `log.(x)` in `_logpdf(d::TuringMvLogNormal, x::AbstractVector{<:Real})`. In a second step one should consider fixing potential type stability problems in `_logpdf(d::TuringMvLogNormal, x::AbstractMatrix{<:Real})`.

Reply:
Thanks for this. Pre-computing `log.(x)` is important here. I will fix it thanks.

Reply:
The main point I was trying to make is that the best performance can't be achieved by pre-computing `log.(x)` alone. In my benchmarks above, it really requires implementation `h` AND pre-computing `log.(x)`. I can reproduce these findings also with your latest commits (after fixing a newly introduced bug, see below), and implementation `h` still is the fastest one with the fewest allocations (14 instead of18, and 401 instead of 703).

BTW the latest commits introduced a bug in the implementation of the logpdf for vectors. `log.(x)` should be precomputed only if `x` is in the support, the current implementation throws `DomainError`s if `x` is not in the support (see my implementation above).

Reply:
I used your implementation thanks :)